_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,644 @@
+
+Install [cert-manager](https://cert-manager.io/docs/tutorials/getting-started-with-cert-manager-on-google-kubernetes-engine-using-lets-encrypt-for-ingress-ssl/) in your cluster:
+
+```shell
+kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.12.1/cert-manager.yaml
+```
+
+**Note**: If you see a `kubectl` error __The gcp auth plugin has been removed. __, see [https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke](https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke) for details on how to fix it.
+
+Apply the deployment descriptors, from the `kubernetes` folder:
+
+```shell
+./kubectl-apply.sh -f
+```
+
+**Important note**: Keycloak client applications (a monolith or microservices) will fail the startup until Let's Encrypt has issued the certificate and it has been synchronized inside the cluster. In the following section, the issuance process is explained in more detail.
+
+The certificate might take some minutes to be ready, you can check the status by inspecting the object events:
+
+```shell
+kubectl describe certificate keycloak-ssl -n demo
+```
+
+The output should be similar to the following text:
+
+```text
+Events:
+  Type    Reason     Age   From                                       Message
+  ----    ------     ----  ----                                       -------
+  Normal  Issuing    10m   cert-manager-certificates-trigger          Issuing certificate as Secret does not exist
+  Normal  Generated  10m   cert-manager-certificates-key-manager      Stored new private key in temporary Secret resource "keycloak-ssl-rsfkj"
+  Normal  Requested  10m   cert-manager-certificates-request-manager  Created new CertificateRequest resource "keycloak-ssl-mp8s7"
+  Normal  Issuing    10m   cert-manager-certificates-issuing          Issued temporary certificate
+  Normal  Issuing    4m    cert-manager-certificates-issuing          The certificate has been successfully issued
+```
+Also, the ingress service will be updated once the certificate is available. You can check the ingress events with the following command:
+
+```shell
+kubectl describe ingress gateway -n demo
+```
+
+The output should contain the message "certs updated":
+
+```text
+Events:
+  Type     Reason             Age                   From                       Message
+  ----     ------             ----                  ----                       -------
+...
+  Normal   Sync               78s (x2 over 4m33s)  loadbalancer-controller    UrlMap "k8s2-um-ag6evarw-demo-gateway-gijhgcnp" updated
+  Normal   Sync               66s (x9 over 9m6s)   loadbalancer-controller    Scheduled for sync
+  Normal   Sync               25s                  loadbalancer-controller    TargetProxy "k8s2-ts-ag6evarw-demo-gateway-gijhgcnp" certs updated
+```
+Once the cluster is healthy, you can test the deployment by navigating to **http://gateway.\<namespace\>.\<public-ip\>.nip.io** :
```

Comment:
When I first tried to go to `http://gateway.demo.34.111.205.125.nip.io/`, I got an error:

<img width="1003" alt="Screenshot 2023-06-30 at 4 53 25 PM" src="https://github.com/oktadev/okta-blog/assets/17892/51b855fb-86b6-4271-9c95-5480af3b6cab">

Reply:
This is probably because the certificate is still being issued. 

```
$ kubectl describe ingress gateway -n demo

Events:
  Type    Reason     Age    From                                       Message
  ----    ------     ----   ----                                       -------
  Normal  Issuing    13m    cert-manager-certificates-trigger          Issuing certificate as Secret does not exist
  Normal  Generated  13m    cert-manager-certificates-key-manager      Stored new private key in temporary Secret resource "keycloak-ssl-59t4w"
  Normal  Requested  13m    cert-manager-certificates-request-manager  Created new CertificateRequest resource "keycloak-ssl-w67hm"
  Normal  Issuing    13m    cert-manager-certificates-issuing          Issued temporary certificate
  Normal  Issuing    7m32s  cert-manager-certificates-issuing          The certificate has been successfully issued
```

Reply:
Added the note.